### ISSUE TITLE
Discrepancy between tf.node.decodePng and tf.browser.fromPixels

### ISSUE THREAD

**Issue by thekevinscott** [tensorflow/tfjs#6627]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow.js): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Browser and MacOS 12.4
- TensorFlow.js installed from (npm or script link): npm
- TensorFlow.js version (use command below): `3.18.0`
- Browser version: Chrome Version 103.0.5060.114 (Official Build) (x86_64)

**Describe the current behavior**
Reading in an image on Node vs the browser produces different tensor values.

**Describe the expected behavior**

I expect the tensors produced from reading the image to be identical across both the browser and Node.

**Standalone code to reproduce the issue**

Using this 4x4 image as an example:
![flower-small](https://user-images.githubusercontent.com/151596/179236399-ec2ce184-a248-4dfe-8140-d4b16e90e8e7.png)

Browser code:

```
import * as tf from '@tensorflow/tfjs';
import imgPath from "./flower-small.png";
const img = new Image();
img.src = imgPath;
img.onload = () => {
  const imageTensor = tf.browser.fromPixels(img);
  imageTensor.print();
};
```

Produces the following output:

```
Tensor
    [[[78 , 108, 53],
      [241, 95 , 52],
      [189, 71 , 27],
      [126, 33 , 1 ]],

     [[214, 117, 67],
      [217, 51 , 26],
      [227, 56 , 28],
      [230, 74 , 33]],

     [[209, 71 , 39],
      [222, 64 , 0 ],
      [240, 82 , 36],
      [215, 65 , 31]],

     [[18 , 86 , 57],
      [194, 121, 92],
      [175, 90 , 37],
      [119, 74 , 24]]] 
```

Node code:

```
const file = fs.readFileSync('./flower-small.png');
const image = tf.node.decodeImage(file, 3);
image.print();
```

Produces the following output:

```
Tensor
    [[[79 , 108, 56],
      [241, 95 , 55],
      [188, 73 , 33],
      [125, 38 , 6 ]],

     [[213, 116, 69],
      [216, 54 , 32],
      [226, 59 , 34],
      [229, 75 , 38]],

     [[208, 73 , 43],
      [221, 66 , 0 ],
      [240, 83 , 41],
      [214, 67 , 36]],

     [[25 , 87 , 60],
      [193, 120, 92],
      [174, 90 , 42],
      [118, 75 , 30]]]
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

[This issue appears to have been brought up before](https://github.com/tensorflow/tfjs/issues/4843#issuecomment-805067717), but with different results, and does not appear to have been addressed in that issue thread.

**Comment by thekevinscott** [comment]
An update on the above.

It seems the discrepancy is not between Node and the browser, but between different strategies of loading images in the browser.

Here's some code sandboxes:

- Node: https://codesandbox.io/s/tfjs-node-image-tensor-6k53yw?file=/index.js
- Browser, load via `<img />` tag: https://codesandbox.io/s/tfjs-browser-image-tensor-identical-to-node-zc3ysr?file=/src/index.js
- Browser, load via `import` path: https://codesandbox.io/s/tfjs-browser-image-tensor-troubleshooting-otctpt?file=/index.js

The first two links produce the same tensor. The third produces a different tensor.

I've no idea why this would be. Does this maybe have something to do with the bundler bundling the code? (It's using Parcel.)

**Comment by danwexler** [comment]
So, the tl;dr is to avoid using HTMLImageElement to load images in the
browser for TFJS, at least if you want the same results in node?

**Comment by thekevinscott** [comment]
To be honest, I have no idea. I ran some more tests that confused things more.

[Here's a browser code sandbox that loads an image from picsum](https://codesandbox.io/s/tfjs-browser-image-tensor-picsum-5u9oc0) in both JPEG and PNG formats. It loads the image over HTTP (only jpeg), imports it via Parcel (both jpeg and png), and pulls it from an HTMLImageElement (both jpeg and png).

All tensor outputs are the same.

[Here's a Node code sandbox that does the same thing in Node](https://codesandbox.io/s/tfjs-browser-image-tensor-picsum-5u9oc0). Strangely, here, the png example is identical, but the **jpeg** produces a different tensor.

TLDR? I really have no clue. The first examples indicate that there may be a discrepancy in terms of how images are loaded, but this set of examples indicates it may have something to do with the image itself.

**Comment by danwexler** [comment]
FWIW, avoiding HTMLImageElement is a good idea because it doesn't work in
web workers either.

**Comment by lina128** [comment]
Hi, the discrepancy comes from different mechanism of rendering an image. In some environments, images are rendered on CPU using CPU rendering engine, in other environments, images are rendered on GPU using GPU rendering engine. Take a look at this [discussion](https://github.com/tensorflow/tfjs/issues/5482) for more pointers. The discrepancy should be large, in a sense that they should both look to you as same image. If we feed this image to an image model, it shouldn't generate significantly different results (e.g. different class labels, etc.).

**Comment by thekevinscott** [comment]
Thank you for the explanation and the links to the other two issues, @lina128. That makes sense to me. Indeed, the images do not appear to be visually different, just the tensor representations.

I opened this issue because I was seeing discrepancies in my model's outputs, but on further investigation I found this was due to a non-TFJS-related bug.

I'll go ahead and close this issue.

**Comment by vladmandic** [comment]
@thekevinscott you might want to take a look at <https://github.com/vladmandic/face-api/issues/114#issuecomment-1202637688>

**Comment by thekevinscott** [comment]
Thanks for the link, @vladmandic, that's a helpful thread to read. I wasn't aware of the lossy nature of JPEG, that would explain the discrepancies I saw in my second comment.

However, isn't PNG a lossless format? Shouldn't that imply that the pixels would be identical across platforms? Or no?

**Comment by vladmandic** [comment]
> However, isn't PNG a lossless format? 

True  

But if you're using canvas, it still leaves browser rendering (browser doesn't just decode and present pixel data as-is) to your color space  

And in case of WebGL backend, also some minute pixel shifts as coordinate space is not identical  

None of that is really visible to human eye, I just wanted to share the thread that shows how minute input changes can result in different model outputs

**Comment by thekevinscott** [comment]
Gotcha. I appreciate the information! TIL!
